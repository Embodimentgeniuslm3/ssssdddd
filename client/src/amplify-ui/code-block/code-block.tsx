import {Component, Host, h, Prop, State} from "@stencil/core";
import {
  codeBlockStyle,
  lineCountStyle,
  oneLineStyle,
  copyButtonStyle,
  slotContainerStyle,
} from "./code-block.style";
import copy from "copy-to-clipboard";

const COPY = "copy";
const COPIED = "copied";
const FAILED = "failed to copy";

let alwaysRecompute = 0;

@Component({tag: "amplify-code-block", shadow: false})
export class AmplifyCodeBlock {
  /** the number of lines of the code block */
  @Prop() readonly lineCount?: string;
  /** what language are we displaying */
  @Prop() readonly language?: string;

  @State() copyMessage: typeof COPY | typeof COPIED | typeof FAILED = "copy";

  element?: HTMLDivElement;

  setElement = (ref: HTMLDivElement | undefined) => (this.element = ref);

  copyToClipboard = () => {
    if (this.element && this.element.textContent) {
      copy(this.element.textContent);
      this.copyMessage = COPIED;
    } else {
      this.copyMessage = FAILED;
    }
    setTimeout(() => {
      this.copyMessage = COPY;
    }, 1500);
  };

  copyButton = () => {
    if (this.language && this.language === "console") return;

    return (
      <button class={copyButtonStyle} onClick={this.copyToClipboard}>
        <span>{this.copyMessage}</span>
      </button>
    );
  };

  render() {
    const parsed = this.lineCount && parseInt(this.lineCount);
    const parsedLineCount = parsed && parsed > 1 ? parsed : undefined;

    return (
      <Host
        class={{
          [codeBlockStyle]: true,
          [oneLineStyle]: parsedLineCount === undefined,
        }}
      >
        {parsedLineCount && (
          <div class={lineCountStyle}>
            <div>
              {this.lineCount &&
                new Array(parsedLineCount)
                  .fill(null)
                  .map((_, i) => <span>{String(i + 1)}</span>)}
            </div>
          </div>
        )}
        <div
          key={String(alwaysRecompute++)}
          ref={this.setElement}
          class={slotContainerStyle}
        >
          <slot name="content" />
        </div>
        {this.copyButton()}
      </Host>
    );
  }
}
