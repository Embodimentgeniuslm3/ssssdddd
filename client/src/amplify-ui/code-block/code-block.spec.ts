import {AmplifyCodeBlock} from "./code-block";
import {newSpecPage} from "@stencil/core/testing";

describe("amplify-code-block", () => {
  describe("Component logic", () => {
    let codeBlock: AmplifyCodeBlock;
    beforeEach(() => (codeBlock = new AmplifyCodeBlock()));

    it("should init `parsedLineCount` as `undefined`", () =>
      expect(codeBlock.parsedLineCount).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyCodeBlock],
            html: `<amplify-code-block />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });

  describe("Copy button", () => {
    it("is visible for languages other than console", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyCodeBlock],
            html: `<amplify-code-block language="javascript" />`,
          })
        ).root,
      ).toMatchSnapshot();
    });

    it("is not visible if the language is set to console", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyCodeBlock],
            html: `<amplify-code-block language="console" />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
