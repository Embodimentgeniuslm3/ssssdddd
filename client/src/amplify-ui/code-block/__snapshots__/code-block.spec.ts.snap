// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-code-block Copy button is not visible if the language is set to console 1`] = `
<amplify-code-block class="css-1lj8awz css-75wbz6" language="console">
  <!---->
  <div class="css-khgmir"></div>
</amplify-code-block>
`;

exports[`amplify-code-block Copy button is visible for languages other than console 1`] = `
<amplify-code-block class="css-1lj8awz css-75wbz6" language="javascript">
  <!---->
  <div class="css-khgmir"></div>
  <button class="css-16vr62x">
    <span>
      copy
    </span>
  </button>
</amplify-code-block>
`;

exports[`amplify-code-block Render logic should render 1`] = `
<amplify-code-block class="css-1lj8awz css-75wbz6">
  <!---->
  <div class="css-khgmir"></div>
  <button class="css-16vr62x">
    <span>
      copy
    </span>
  </button>
</amplify-code-block>
`;
